data-prepper-core/src/main/java/com/amazon/dataprepper/parser/PipelineParser.java

```diff
@@ -119,6 +123,19 @@ private void buildPipelineFromConfiguration(
 
     }
 
+    private List<Prepper> newPreppers(final PluginSetting pluginSetting) {
```

Comment:
This feels like a leak of plugin loading responsibilities of the PluginFactory into the PIpelineParser. I would expect the PluginFactory to support this. The PipelineParser is responsible for instantiating plugins based on the class. 

If we know there are going to be class types that may require multiple instances. I would recommend the following change to the PluginFactory. This would support only the PluginFactory instantiating classes and eliminate the PipelineParser from knowing about how the classes are defined. 

```
public interface PluginFactory {
    /**
     * Loads a new instance of a plugin.
     *
     * @param baseClass The class type that the plugin is supporting.
     * @param pluginSetting The {@link PluginSetting} to configure this plugin
     * @param <T> The type
     * @return A new instance of your plugin, configured
     * @since 1.2
     */
    <T> T loadPlugin(final Class<T> baseClass, final PluginSetting pluginSetting);

    /**
     * Loads multiple instances of a plugin based on the configuration.
     *
     * @param baseClass The class type that the plugin is supporting.
     * @param pluginSetting The {@link PluginSetting} to configure this plugin
     * @param <T> The type
     * @return A collection of new instances of your plugin, configured
     * @since 1.2
     */
    <T> LIst<T> loadPlugin(final Class<T> baseClass, final PluginSetting pluginSetting);
}
```

Reply:
I am open to other names for the multiple instances loader as well. I struggled with this a little bit.

Reply:
I pushed a change with `loadProcessWorkerPlugins`, but I think we are moving some special logic into the `PluginFactory` itself.

I'd like to further decouple these by using a `Function<Class, Integer>` to determine the total number of instances to create.
We can alter the signature of `PluginFactory` by adding this method:

```
    /**
     * Loads a specified number of plugin instances. The total number of instances is provided
     * by the numberOfInstancesFunction.
     *
     * @param baseClass The class type that the plugin is supporting.
     * @param pluginSetting The {@link PluginSetting} to configure this plugin
     * @param numberOfInstancesFunction A {@link Function} which takes as input a {@link Class}
     *                                  and returns an {@link Integer}. The input class is the actual
     *                                  class plugin class instance. The returned integer value is
     *                                  the total count of instances to create.
     * @param <T> The type
     * @return One or more new instances of the plugin
     * @since 1.2
     */
    <T> List<T> loadPlugins(
            final Class<T> baseClass, final PluginSetting pluginSetting,
            final Function<Class<T>, Integer> numberOfInstancesFunction);
```

And then `newPreppers` looks like:

```
    private List<Prepper> newPreppers(final PluginSetting pluginSetting) {
        return pluginFactory.loadPlugins(Prepper.class, pluginSetting,
                actualClass -> actualClass.isAnnotationPresent(SingleThread.class) ?
                pluginSetting.getNumberOfProcessWorkers() :
                1);
    }
```

In this way, the `PluginFactory` does not provide an extra `getPlugin` method and is still responsible for loading the plugins. But, the caller (in this case `PipelineParser`) will determine the total count by introspecting the class.